### PR TITLE
Snowflake: Complete PATTERN grammar

### LINKED ISSUE #2783: Add complete Row Pattern Recognition grammar to Snowflake
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

In December 2016, the International Organization for Standardization (ISO) released a new version of the SQL standard which includes Row Pattern Recognition in SQL (ISO/IEC 9075:2016 or ANSI X3.135)

Snowflake supports this feature (see [Snowflake documentation](https://docs.snowflake.com/en/sql-reference/constructs/match_recognize.html)) as well as some other ANSI compliant SQL engines (e.g. [Trino](https://trino.io/docs/current/sql/match-recognize.html) (f.k.a. PrestoSQL)).

### Use case

It would be nice if SQLFluff supports this syntax.

### Dialect

ANSI
Snowflake

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by jpy-git**
### Brief summary of the change made

Follows on from #2781. Closes #2783. Implements complete PATTERN grammar in Snowflake.

N.B. as mentioned in the linked issue, `MATCH_RECOGNIZE` is in the ANSI standard and will likely be implemented in other dialects in the future. Currently its only implemented for Snowflake and Trino (which we don't currently support) so we can leave it in the Snowflake dialect for now. Should be a pretty straight copy from snowflake when needed.

### Are there any other side effects of this change that we should be aware of?
No

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_snowflake_keywords.py [hunk 1]
```diff
@@ -49,6 +49,7 @@
 LIKE
 LOCALTIME
 LOCALTIMESTAMP
+MATCH_RECOGNIZE
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_snowflake_keywords.py:52]: Had to move this to reserved keywords to stop it from being captured as an alias

File: test/fixtures/dialects/snowflake/snowflake_match_recognize.sql [hunk 2]
```diff
@@ -18,7 +19,66 @@ select * from stock_price_history
   )
 order by company, match_number;
 
--- With more complicated pattern
+select price_date, match_number, msq, price, cl from
+  (select * from stock_price_history where company='ABCD') match_recognize(
```
Comment (jpy-git) [inline test/fixtures/dialects/snowflake/snowflake_match_recognize.sql:23]: This was the query where match_recognize kept getting matched as an alias for the subquery. I tried adding it to some of the terminator grammars to no avail. its unlikely someone is using `match_recognize` as a column name so we should be pretty safe to just make match_recognize reserved for now

File: test/fixtures/dialects/snowflake/snowflake_match_recognize.sql [hunk 3]
```diff
@@ -38,3 +98,64 @@ select * from stock_price_history
...
+    measures
+      match_number() as match_number,
+      first(price_date) as start_date,
+      last(price_date) as end_date,
+      count(*) as rows_in_sequence,
+      count(row_with_price_decrease.*) as num_decreases,
+      count(row_with_price_increase.*) as num_increases
+    one row per match
+    after match skip to last row_with_price_increase
+    pattern(^ ( S1 | S2* )? S3 PERMUTE(S4+, S5*?) {- S6 -}+ $)
```
Comment (jpy-git) [inline test/fixtures/dialects/snowflake/snowflake_match_recognize.sql:116]: example of robust pattern grammar 💪

File: test/fixtures/dialects/snowflake/snowflake_match_recognize.sql [hunk 4]
```diff
@@ -38,3 +98,64 @@ select * from stock_price_history
...
+    measures
+      match_number() as match_number,
+      first(price_date) as start_date,
+      last(price_date) as end_date,
+      count(*) as rows_in_sequence,
+      count(row_with_price_decrease.*) as num_decreases,
+      count(row_with_price_increase.*) as num_increases
+    one row per match
+    after match skip to last row_with_price_increase
+    pattern((A {- B+ C+ -} D+))
```
Comment (jpy-git) [inline test/fixtures/dialects/snowflake/snowflake_match_recognize.sql:136]: some nested pattern grammar to demo recursion

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 5]
```diff
@@ -265,17 +277,66 @@
         Ref("QuestionMarkSegment", optional=True),
         allow_gaps=False,
     ),
+    PatternSymbolGrammar=Sequence(
+        Ref("SingleIdentifierGrammar"),
+        Ref("PatternQuantifierGrammar", optional=True),
+        allow_gaps=False,
+    ),
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_snowflake.py:284]: Any individual symbol can have a quantifier

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 6]
```diff
@@ -265,17 +277,66 @@
...
-                Ref("SingleIdentifierGrammar"),
-                Ref("PatternQuantifierGrammar", optional=True),
-                allow_gaps=False,
+        OneOf(
+            AnyNumberOf(
+                Ref("PatternOperatorGrammar"),
+            ),
+            Delimited(
+                Ref("PatternOperatorGrammar"),
+                delimiter=Ref("BitwiseOrSegment"),
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_snowflake.py:339]: this represents either space or `|` separated operators/symbols

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 7]
```diff
@@ -265,17 +277,66 @@
...
+                            Ref("PatternOperatorGrammar"),
+                        ),
+                        Delimited(
+                            Ref("PatternOperatorGrammar"),
+                            delimiter=Ref("BitwiseOrSegment"),
+                        ),
+                    ),
+                    bracket_type="exclude",
+                    bracket_pairs_set="bracket_pairs",
+                ),
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_snowflake.py:301]: Exclude brackets.
note the recursion of `PatternOperatorGrammar` much like in the expression grammar

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 8]
```diff
@@ -265,17 +277,66 @@
...
+                    OneOf(
+                        AnyNumberOf(
+                            Ref("PatternOperatorGrammar"),
+                        ),
+                        Delimited(
+                            Ref("PatternOperatorGrammar"),
+                            delimiter=Ref("BitwiseOrSegment"),
+                        ),
+                    ),
+                ),
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_snowflake.py:312]: standard brackets

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 9]
```diff
@@ -265,17 +277,66 @@
...
+                Sequence(
+                    "PERMUTE",
+                    Bracketed(
+                        Delimited(
+                            Ref("PatternSymbolGrammar"),
+                        ),
+                    ),
+                ),
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_snowflake.py:320]: permute function

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 10]
```diff
@@ -265,17 +277,66 @@
...
+            ),
+            # Operators can also be followed by a quantifier.
+            Ref("PatternQuantifierGrammar", optional=True),
+            allow_gaps=False,
```
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_snowflake.py:324]: The above 3 operators can also have a quantifier

**Review by tunetheweb — APPROVED** [review]
FYI, you can also prevent an unreserved keyword matching as an alias by explicitly excluding it in `AliasExpressionSegment`:

```py
class AliasExpressionSegment(BaseSegment):
    """A reference to an object with an `AS` clause.

    The optional AS keyword allows both implicit and explicit aliasing.
    """

    type = "alias_expression"
    match_grammar = Sequence(
        Ref.keyword("AS", optional=True),
        OneOf(
            Sequence(
                Ref("SingleIdentifierGrammar"),
                # Column alias in VALUES clause
                Bracketed(Ref("SingleIdentifierListSegment"), optional=True),
            ),
            Ref("SingleQuotedIdentifierSegment"),
            excludes="MATCH_RECOGNIZE",
        ),
    )
```

But I agree it's unlikely to be an alias so not worth the complexity.

**Pull request merged**